### ISSUE TITLE
ValueError when trying to reassign an index column

### ISSUE THREAD

**Issue by multimeric** [dask/dask#9380]
**What happened**:

The following error:
```
ValueError: The columns in the computed data do not match the columns in the provided metadata
  Extra:   [0]
  Missing: ['index']
```

**What you expected to happen**:
No error in this situation, it should compute as expected.

**Minimal Complete Verifiable Example**:
This is highly contrived, because I spent a long time reducing this error to a minimal state from my real world workflow. The `shuffle` method call here is used to represent any more complex operation like a hash join, which is more common in real usage.
```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame({
    "uniprot": ["P42684", "P42684", "P42684"]
}, index=pd.Series(["1:179108338:G>C", "1:179109432:C>T", "1:179126409:G>C"], name="index"))
ddf = dd.from_pandas(df, npartitions=2)

uniprots = ddf.uniprot.unique()
x = uniprots.to_frame().assign(index = uniprots).set_index("index")
x.shuffle(x.index).compute()
```

**Anything else we need to know?**:

**Environment**:

- Dask version: `2022.8.0`
- Python version: `3.9.5`
- Operating System: Ubuntu 21.10
- Install method (conda, pip, source): pip

**Comment by multimeric** [comment]
Okay so on further inspection I think this issue relates to the fact that the fact that dask isn't aware that a column has been converted into an index, which can be seen by looking at the meta. In the above example, if we `print(x._meta)`, we get:
```
Empty DataFrame
Columns: [uniprot, index]
Index: []
```

So even though we've explicitly `set_index("index")`, dask still thinks that "index" is a column, and that the actual index has no name, which is not correct.

Another example which I believe demonstrates this same behaviour is here:

```python
import pandas as pd
import dask.dataframe as dd
from dask.dataframe.utils import make_meta
from dask.dataframe.multi import hash_join

df = dd.from_pandas(pd.DataFrame({
    "a": range(10),
    "b": range(10),
}), npartitions=2)

df2 = df.map_partitions(lambda df: df.set_index("a"), meta={"b": int})
print(df2._meta)
hash_join(df2, df2.index, df, df.index).compute()
```
```
Empty DataFrame
Columns: [b]
Index: []

ValueError: The columns in the computed data do not match the columns in the provided metadata
  Extra:   ['a']
  Missing: [0]
```

We set the index of `df2` to `a` implicitly, and we can see that `dask` is not aware of this. Then when we try to hash join (which again, is a simplification of a real life operation like a normal join), dask has issues with the mislabelled index.

**Comment by multimeric** [comment]
And here's another example that doesn't use `map` at all:
```python
import pandas as pd
import dask.dataframe as dd
from dask.dataframe.multi import hash_join

df = dd.from_pandas(pd.DataFrame({
    "a": range(10),
    "b": range(10),
}), npartitions=2)
dfa = df.set_index("a")
dfb = df.set_index("b")
x = hash_join(dfa, dfa.index, dfb, dfb.index)
y = hash_join(x, x.index, dfa, dfa.index).compute()
```
```
ValueError: The columns in the computed data do not match the columns in the provided metadata
  Extra:   [0]
  Missing: ['a']
```

**Comment by jsignell** [comment]
I think there are a couple different things going on here. In the first example, Dask is trying to be a bit too clever. It sees that the name that the column that you are trying to assign to index matches `ddf.index` and so the `set_index` bails early.  If you don't rename the column, and just set index with it as it is originally named things work fine:

```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame(
    {"uniprot": ["P42684", "P42684", "P42684"]},
    index=pd.Series(
        ["1:179108338:G>C", "1:179109432:C>T", "1:179126409:G>C"], name="index"
    ),
)
ddf = dd.from_pandas(df, npartitions=2)

uniprots = ddf.uniprot.unique()
x = uniprots.to_frame().set_index("uniprot")
x.shuffle(x.index).compute()
```

So that is one little bug. 

The rest of the examples that you provide have the intended behavior. In Dask when you are doing custom functionality you often need to set the meta to show what the output will look like.  So for the second example, you should use a pandas object that has the right index name and dtype. Something like:

```python
import pandas as pd
import dask.dataframe as dd
from dask.dataframe.utils import make_meta
from dask.dataframe.multi import hash_join

df = dd.from_pandas(pd.DataFrame({
    "a": range(10),
    "b": range(10),
}), npartitions=2)

df2 = df.map_partitions(lambda df: df.set_index("a"), meta=df.head().set_index("a"))
print(df2._meta)
hash_join(df2, df2.index, df, df.index).compute()
```

For the third case I agree that it looks like there is maybe a little bug in hash join where Dask doesn't have the right expectation of the index name.

**Comment by multimeric** [comment]
Thanks for the response. I agree that issue 2 is user error, I forget why I included it exactly. Possibly I was trying to make a point about dask needing a correctly named index which means it's not trivial to rename the index in point 1, because it will throw off all the downstream metadata unless you change it all.

**Comment by jsignell** [comment]
For the first issue. We recently added a warning, but maybe that was the wrong approach. Maybe we should stop bailing early in that case or maybe we should raise an error. I think there is a little thought left to do there. 

For the hash_join though it's probably a straightforward fix, so you are welcome to open a pull request if you feel so inclined.

**Comment by multimeric** [comment]
Yes so I've noticed this warning (`UserWarning: New index has same name as existing, this is a no-op.`), but I didn't interpret correctly. I guess if you keep it, I would rather it be more clear, ie `New index has same name as existing, index assignment will be skipped`. But also, it must still be doing *something*, right? Or the metadata error wouldn't happen at all. Is it changing the metadata but not the index itself, perhaps?

Anyway, imo I would rather it not bail easy, and just does the operation even if it may seem redundant. I guess the warning could work the other way around, such as "the index has the same name, this may be redundant".

**Comment by jsignell** [comment]
>  But also, it must still be doing _something_, right? Or the metadata error wouldn't happen at all. 

That's a good point, but it's not. You get the same metadata error even without the `set_index` in the first example. If you change the last line to `x.shuffle(x["index"]).compute()` it works though. I just took a little look and gross as it is _I think_ this matches pandas behavior. See if you do `x.compute().index` you get a `RangeIndex` and if you do `x.compute()["index"]` you get the column named index. 

> Anyway, imo I would rather it not bail early, and just do the operation even if it may seem redundant. 

I tend to agree. Probably better to have a fixed cost rather than try to be clever if we don't have enough information to go on.

**Comment by multimeric** [comment]
Okay so in my example I create a column named index, then call `set_index()` which does nothing, then try to shuffle it using the old index. Why does this then fail?

**Comment by jsignell** [comment]
Hmm. There must be something wrong with the meta in shuffle for the case where a dd.Index object is passed. 

```python

import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame(
    {"uniprot": ["P42684", "P42684", "P42684"]},
    index=pd.Series(
        ["1:179108338:G>C", "1:179109432:C>T", "1:179126409:G>C"], name="index"
    ),
)
ddf = dd.from_pandas(df, npartitions=2)

uniprots = ddf.uniprot.unique()
x = uniprots.to_frame()
x.shuffle("index").compute()  # this works
x.shuffle(x.index).compute()  # this does not
```

**Comment by jorloplaz** [comment]
I mentioned this discussion in the above PR. My general view on this is the following:

- **A named index in a Pandas dataframe is actually a column in the original data** which, for convenience, is acting TEMPORARILY as the index.
- For this reason, Pandas by default uses an UNNAMED range index. **It is the fact that it is unnamed that makes clear this index is NOT a column of your data**, but only a convenient way to index by row number (i.e., `[0, ..., len(df)-1]`).
- Dask should follow this same convention: **a named index can at any moment be reset to be again a column (preserving its name), while an unnamed index is just accessory and should be disposed when reset**.

So, given that the above PR should prevent you from having any Dask dataframe with 2 columns with the same name, the potentially conflicting cases are the following ones:

- `ddf.reset_index()` when `ddf.index = None`: this should NOT create a column called `'index'` (because that was never a column in the original data!), so it should be dropped. Explicitly calling `ddf.reset_index(drop=False)` should raise an error.
- `ddf.set_index(series)`: with the above PR it is still possible to do things like `ddf.set_index(ddf['x'])`, which is convenient in some particular cases, such as when joining by index (so columns need to be first set as indexes). However, note that chaining `ddf.set_index(ddf['x']).reset_index(drop=False)` will fail, because that would make you have 2 columns named `'x'`.

In summary, once the PR is accepted, you have this situation:

1. **Your original data CANNOT have any columns named equally**. In the particular case of a Pandas dataframe (i.e., you build your Dask dataframe `ddf` with `from_pandas`), this ALSO includes the index; **the index of the Pandas frame CANNOT have the same name as any column**, to begin with.
2. Once you have `ddf`, you are indeed allowed to index by a column (or by a series derived from a column which preserves its name, such as `ddf['x'] + 1`), but this is something that you are doing explicitly, who knows why...
3. Whatever yourreasons are, this index CANNOT be reset to a column unless you rename it previously. It makes sense to be so strict: 
- if you indexed with a column via `ddf.set_index(ddf['x'])` instead of `ddf.set_index('x')`, `'x'` is still a column, so there's no point in calling `reset_index(drop=False)`.
- if you indexed with something derived from a column (e.g., `ddf.set_index(ddf['x']+1)`, and you want that to become a column, it forces you to name that new column with a suitablename, either directly via `ddf.index = 'xPlus1'`, or indirectly via `ddf.set_index((ddf['x']+1).rename('xPlus1'))`.

Hope you find this sensible!
